### PR TITLE
feat: add invalidate endpoint

### PR DESCRIPTION
**Pull request by EslamHiko**
- [ ] This is a **bugfix**
- [x] This is a **feature**
- [ ] This is a **code refactor**
- [x] This is a **test update**
- [ ] This is a **docs update**
- [ ] This is a **metadata update**

### For Bugs and Features; did you add new tests?
yes

complete: https://github.com/webpack/webpack-dev-server/pull/1546
### Breaking Changes
no

### CODE REVIEW

File: test/server/utils/routes.test.js [hunk 1]
```diff
@@ -64,6 +64,10 @@ describe('routes util', () => {
     });
   });
 
+  it('GET request to invalidate endpoint', (done) => {
+    req.get('/invalidate').expect(200, done);
+  });
```
Comment (evilebottnawi) [inline test/server/utils/routes.test.js:69]: Will be great to test what we really do invalidate - get stats before do request and do get stats again

Comment (EslamHiko) [reply test/server/utils/routes.test.js:69]: I can't detect any changes in `server.middleware`

Comment (EslamHiko) [reply test/server/utils/routes.test.js:69]: and `server.middleware.context.stats` is always `undefined`

Comment (evilebottnawi) [reply test/server/utils/routes.test.js:69]: Strange, should be no

Comment (EslamHiko) [reply test/server/utils/routes.test.js:69]: I've thought about an Idea maybe we can create an option `invalidateCallback` and we create a function that creates a file then delete after confirming that the file exists.

Comment (evilebottnawi) [reply test/server/utils/routes.test.js:69]: Bad idea, maybe we need fix something in tests, need debug

File: lib/Server.js [hunk 2]
```diff
@@ -118,7 +118,7 @@ class Server {
     this.setupDevMiddleware();
 
     // set express routes
-    routes(this.app, this.middleware, this.options);
```
Comment (hiroppy) [inline lib/Server.js:121]: Why did you change this?

Comment (EslamHiko) [reply lib/Server.js:121]: @hiroppy To pass the server, so I can call server.invalidate()

Comment (hiroppy) [reply lib/Server.js:121]: I'm -1 on this change because if you pass `this`, we can't understand what variables are used clearly.

Comment (EslamHiko) [reply lib/Server.js:121]: @hiroppy I tried `(this.app, this.middleware, this.options, this.invalidate)` and called `invalidate()` only but this resulted in `500 "Internal Server Error"` so what about changing `this` into `server` for better reading ex: 
```
const server = this;
routes(server);
```

Comment (EslamHiko) [reply lib/Server.js:121]: one thing that worked was moving 
```javascript
invalidate(callback) {
    if (this.middleware) {
      this.middleware.invalidate(callback);
    }
  }
``` 
to `routes.js` and make it only accessable via the api call, but that would require adding an option `invalidateCallBack` so we can pass it ex: 
```javascript
    this.invalidateCallBack = this.options.invalidateCallBack;
    // set express routes
    routes(this.app, this.middleware, this.options, this.invalidateCallBack);
```
and in `routes.js` 
```javascript
function routes(app, middleware, options, invalidateCallBack) {
  app.get('/invalidate', (_req, res) => {
    if(middleware)
      middleware.invalidate(invalidateCallBack)
    res.end();
  });
}
```

**Comment by evilebottnawi** [comment]
@hiroppy I think it is normal to pass `this` in some cases, like webpack pass `compiler` and `compilation` to hooks and utils functions

**Comment by hiroppy** [comment]
I got it but it's hard to read for me like a black box.

**Comment by evilebottnawi** [comment]
@hiroppy I think we can do refactor for utils and code for next branch, we have many bad places that we can improve, maybe using destructuring is good idea in many places

**Comment by hiroppy** [comment]
@evilebottnawi Totally agree with you. I'll improve these codes on the next branch.

**Comment by evilebottnawi** [comment]
@hiroppy So i think we can merge it, we should fix the problem https://github.com/webpack/webpack-dev-server/issues/1720 (we can't remove inline while not fix it + it will good for next) and do release

File: lib/utils/routes.js [hunk 3]
```diff
@@ -30,6 +34,11 @@ function routes(app, middleware, options) {
     createReadStream(join(clientBasePath, 'live.html')).pipe(res);
   });
 
+  app.get('/invalidate', (_req, res) => {
```
Comment (sokra) [inline lib/utils/routes.js:37]: This should better be prefix with `webpack-dev-server` maybe `/webpack-dev-server/invalidate`. Otherwise `/invalidate` is unusable for the user and apps using `/invalidate` in their router break.
